File: app/src/main/java/com/mthaler/knittings/PhotoGalleryFragment.java
package com.mthaler.knittings;

import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.support.v4.app.Fragment;
import android.support.v4.content.FileProvider;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.GridView;
import java.io.File;
import java.util.List;

public class PhotoGalleryFragment extends Fragment {

    private static final String CURRENT_PHOTO_PATH = "current_photo_path";
    private static final String KNITTING_ID = "knitting_id";
    private static final int REQUEST_IMAGE_CAPTURE = 0;
    private static final String LOG_TAG = PhotoGalleryFragment.class.getSimpleName();

    private Knitting knitting = null;

    private GridView gridView;
    private File currentPhotoPath;

    public static PhotoGalleryFragment newInstance(Knitting knitting) {
        final PhotoGalleryFragment fragment = new PhotoGalleryFragment();
        Bundle args = new Bundle();
        args.putLong(KNITTING_ID, knitting.getId());
        fragment.setArguments(args);
        Log.d(LOG_TAG, "Created new PhotoGalleryFragment with knitting id: " + knitting.getId());
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        final View v = inflater.inflate(R.layout.fragment_photo_gallery, container, false);

        if (savedInstanceState != null) {
            if (savedInstanceState.containsKey(CURRENT_PHOTO_PATH)) {
                currentPhotoPath = new File(savedInstanceState.getString(CURRENT_PHOTO_PATH));
                Log.d(LOG_TAG, "Set current photo path: " + currentPhotoPath);
            }
            if (savedInstanceState.containsKey(KNITTING_ID)) {
                knitting = KnittingsDataSource.Companion.getInstance(getActivity()).getKnitting(savedInstanceState.getLong(KNITTING_ID));
                Log.d(LOG_TAG, "Set knitting: " + knitting);
            }
        } else {
            final long knittingID = getArguments().getLong(KNITTING_ID);
            knitting = KnittingsDataSource.Companion.getInstance(getActivity()).getKnitting(knittingID);
            Log.d(LOG_TAG, "Set knitting: " + knitting);
        }

        gridView = v.findViewById(R.id.gridView);
        final List<Photo> photos = KnittingsDataSource.Companion.getInstance(getActivity()).getAllPhotos(knitting);
        final GridViewAdapter gridAdapter = new GridViewAdapter(getActivity(), R.layout.grid_item_layout, photos);
        gridView.setAdapter(gridAdapter);
        gridView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            public void onItemClick(AdapterView<?> parent, View v, int position, long id) {
            if (position < gridView.getAdapter().getCount() - 1) {
                final Photo photo = (Photo) parent.getItemAtPosition(position);
                //Create intent
                final Intent intent = new Intent(getActivity(), PhotoActivity.class);
                intent.putExtra(PhotoActivity.Companion.getEXTRA_PHOTO_ID(), photo.getId());
                intent.putExtra(KnittingActivity.Companion.getEXTRA_KNITTING_ID(), knitting.getId());
                Log.d(LOG_TAG, "Created PhotoActivity intent");

                //Start details activity
                startActivity(intent);
            } else {
                final Intent takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
                final File photoFile = KnittingsDataSource.Companion.getInstance(getActivity()).getPhotoFile(knitting);
                currentPhotoPath = photoFile;
                Log.d(LOG_TAG, "Set current photo path: " + currentPhotoPath);
                final PackageManager packageManager = getActivity().getPackageManager();
                boolean canTakePhoto = photoFile != null && takePictureIntent.resolveActivity(packageManager) != null;
                if (canTakePhoto) {
                    Uri uri = FileProvider.getUriForFile(getContext(), "com.mthaler.knittings.fileprovider", photoFile);
                    takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, uri);
                    Log.d(LOG_TAG, "Created take picture intent");
                    startActivityForResult(takePictureIntent, REQUEST_IMAGE_CAPTURE);
                }
            }
            }
        });

        return v;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode != Activity.RESULT_OK) {
            return;
        }
        if (requestCode == REQUEST_IMAGE_CAPTURE) {
            // add photo to database
            Log.d(LOG_TAG, "Received result for take photo intent");
            final int orientation = PictureUtils.INSTANCE.getOrientation(currentPhotoPath.getAbsolutePath());
            final Bitmap preview = PictureUtils.INSTANCE.decodeSampledBitmapFromPath(currentPhotoPath.getAbsolutePath(), 200, 200);
            final Bitmap rotatedPreview = PictureUtils.INSTANCE.rotateBitmap(preview, orientation);
            final Photo photo = KnittingsDataSource.Companion.getInstance(getActivity()).createPhoto(currentPhotoPath, knitting.getId(), rotatedPreview, "");
            Log.d(LOG_TAG, "Created new photo from " + currentPhotoPath + ", knitting id " + knitting.getId());
            // add first photo as default photo
            if (knitting.getDefaultPhoto() == null) {
                Log.d(LOG_TAG, "Set " + photo + " as default photo");
                knitting.setDefaultPhoto(photo);
                KnittingsDataSource.Companion.getInstance(getActivity()).updateKnitting(knitting);
            }
            // update grid view
            final List<Photo> photos = KnittingsDataSource.Companion.getInstance(getActivity()).getAllPhotos(knitting);
            final GridViewAdapter gridAdapter = new GridViewAdapter(getActivity(), R.layout.grid_item_layout, photos);
            gridView.setAdapter(gridAdapter);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        if (knitting != null) {
            final List<Photo> photos = KnittingsDataSource.Companion.getInstance(getActivity()).getAllPhotos(knitting);
            final GridViewAdapter gridAdapter = new GridViewAdapter(getActivity(), R.layout.grid_item_layout, photos);
            gridView.setAdapter(gridAdapter);
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        if (currentPhotoPath != null) {
            outState.putString(CURRENT_PHOTO_PATH, currentPhotoPath.getAbsolutePath());
        }
        if (knitting != null) {
            outState.putLong(KNITTING_ID, knitting.getId());
        }
        super.onSaveInstanceState(outState);
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        if (knitting != null) {
            if (isVisibleToUser) {
                // the fragment became visible because the user selected it in the view pager
                // get current knitting from database
                knitting = KnittingsDataSource.Companion.getInstance(getActivity()).getKnitting(knitting.getId());
            } else {
                // the fragment became invisible because the user selected another tab in the view pager
                // save current knitting to database
                KnittingsDataSource.Companion.getInstance(getActivity()).updateKnitting(knitting);
            }
        }
    }
}

File: app/src/main/kotlin/com/mthaler/knittings/PhotoGalleryFragment.kt
package com.mthaler.knittings

import android.app.Activity
import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.provider.MediaStore
import android.support.v4.app.Fragment
import android.support.v4.content.FileProvider
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.Button
import android.widget.GridView
import org.jetbrains.anko.find
import org.jetbrains.anko.support.v4.*
import android.widget.Toast
import com.mthaler.knittings.model.Knitting
import com.mthaler.knittings.model.Photo
import org.jetbrains.anko.AnkoLogger
import org.jetbrains.anko.debug
import java.io.*

/**
 * A fragment that displays a list of photos using a grid
 */
class PhotoGalleryFragment : Fragment(), AnkoLogger {

    private var knitting: Knitting? = null

    private lateinit var gridView: GridView
    private var currentPhotoPath: File? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val v = inflater.inflate(R.layout.fragment_photo_gallery, container, false)

        if (savedInstanceState != null) {
            if (savedInstanceState.containsKey(CURRENT_PHOTO_PATH)) {
                currentPhotoPath = File(savedInstanceState.getString(CURRENT_PHOTO_PATH)!!)
                debug("Set current photo path: " + currentPhotoPath)
            }
            if (savedInstanceState.containsKey(KNITTING_ID)) {
                knitting = KnittingsDataSource.getInstance(activity).getKnitting(savedInstanceState.getLong(KNITTING_ID))
                debug("Set knitting: " + knitting)
            }
        } else {
            val knittingID = arguments.getLong(KNITTING_ID)
            knitting = KnittingsDataSource.getInstance(activity).getKnitting(knittingID)
            debug("Set knitting: " + knitting)
        }

        gridView = v.findViewById(R.id.gridView)
        val photos = KnittingsDataSource.getInstance(activity).getAllPhotos(knitting!!)
        val gridAdapter = GridViewAdapter(activity, R.layout.grid_item_layout, photos)
        gridView.adapter = gridAdapter
        gridView.onItemClickListener = AdapterView.OnItemClickListener { parent, v, position, id ->
            if (position < gridView.adapter.count - 1) {
                // photo clicked, show photo in photo activity
                val photo = parent.getItemAtPosition(position) as Photo
                startActivity<PhotoActivity>(PhotoActivity.EXTRA_PHOTO_ID to photo.id, KnittingActivity.EXTRA_KNITTING_ID to knitting!!.id)
            } else {
                // take photo icon clicked, ask user if photo should be taken or imported from gallery
                val b = AlertDialog.Builder(activity)
                val layout = inflater.inflate(R.layout.dialog_take_photo, parent, false)
                val buttonTakePhoto = layout.find<Button>(R.id.button_take_photo)
                val buttonImportPhoto = layout.find<Button>(R.id.buttom_import_photo)
                b.setView(layout)
                b.setNegativeButton(R.string.dialog_button_cancel) { diag, i -> diag.dismiss()}
                val d = b.create()
                buttonTakePhoto.setOnClickListener {
                    d.dismiss()
                    val takePictureIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
                    currentPhotoPath = KnittingsDataSource.getInstance(activity).getPhotoFile(knitting!!)
                    debug("Set current photo path: " + currentPhotoPath)
                    val packageManager = activity.packageManager
                    val canTakePhoto = currentPhotoPath != null && takePictureIntent.resolveActivity(packageManager) != null
                    if (canTakePhoto) {
                        val uri = FileProvider.getUriForFile(context, "com.mthaler.knittings.fileprovider", currentPhotoPath)
                        takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, uri)
                        debug("Created take picture intent")
                        startActivityForResult(takePictureIntent, REQUEST_IMAGE_CAPTURE)
                    }
                }
                buttonImportPhoto.setOnClickListener {
                    d.dismiss()
                    currentPhotoPath = KnittingsDataSource.getInstance(activity).getPhotoFile(knitting!!)
                    debug("Set current photo path: " + currentPhotoPath)
                    val photoPickerIntent = Intent(Intent.ACTION_PICK)
                    photoPickerIntent.type = "image/*"
                    startActivityForResult(photoPickerIntent, REQUEST_IMAGE_IMPORT)
                }
                d.show()
            }
        }

        return v
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode != Activity.RESULT_OK) {
            return
        }
        if (requestCode == REQUEST_IMAGE_CAPTURE) {
            // add photo to database
            debug("Received result for take photo intent")
            val orientation = PictureUtils.getOrientation(currentPhotoPath!!.absolutePath)
            val preview = PictureUtils.decodeSampledBitmapFromPath(currentPhotoPath!!.absolutePath, 200, 200)
            val rotatedPreview = PictureUtils.rotateBitmap(preview, orientation)
            val photo = KnittingsDataSource.getInstance(activity).createPhoto(currentPhotoPath!!, knitting!!.id, rotatedPreview, "")
            debug("Created new photo from " + currentPhotoPath + ", knitting id " + knitting?.id)
            // add first photo as default photo
            if (knitting!!.defaultPhoto == null) {
                debug("Set $photo as default photo")
                knitting = knitting?.copy(defaultPhoto = photo)
                KnittingsDataSource.getInstance(activity).updateKnitting(knitting!!)
            }
            // update grid view
            val photos = KnittingsDataSource.getInstance(activity).getAllPhotos(knitting!!)
            val gridAdapter = GridViewAdapter(activity, R.layout.grid_item_layout, photos)
            gridView.adapter = gridAdapter
        } else if (requestCode == REQUEST_IMAGE_IMPORT) {
            try {
                debug("Received result for import photo intent")
                val imageUri = data!!.data
                PictureUtils.copy(imageUri, currentPhotoPath!!, activity)
                val orientation = PictureUtils.getOrientation(currentPhotoPath!!.absolutePath)
                val preview = PictureUtils.decodeSampledBitmapFromPath(currentPhotoPath!!.absolutePath, 200, 200)
                val rotatedPreview = PictureUtils.rotateBitmap(preview, orientation)
                val photo = KnittingsDataSource.getInstance(activity).createPhoto(currentPhotoPath!!, knitting!!.id, rotatedPreview, "")
                debug("Created new photo from " + currentPhotoPath + ", knitting id " + knitting?.id)
                // add first photo as default photo
                if (knitting!!.defaultPhoto == null) {
                    debug("Set $photo as default photo")
                    knitting = knitting?.copy(defaultPhoto = photo)
                    KnittingsDataSource.getInstance(activity).updateKnitting(knitting!!)
                }
                // update grid view
                val photos = KnittingsDataSource.getInstance(activity).getAllPhotos(knitting!!)
                val gridAdapter = GridViewAdapter(activity, R.layout.grid_item_layout, photos)
                gridView.adapter = gridAdapter
            } catch (e: FileNotFoundException) {
                e.printStackTrace()
                Toast.makeText(activity, "Something went wrong", Toast.LENGTH_LONG).show()
            }


        }
    }

    override fun onResume() {
        super.onResume()
        if (knitting != null) {
            val photos = KnittingsDataSource.getInstance(activity).getAllPhotos(knitting!!)
            val gridAdapter = GridViewAdapter(activity, R.layout.grid_item_layout, photos)
            gridView.adapter = gridAdapter
        }
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        if (currentPhotoPath != null) {
            outState!!.putString(CURRENT_PHOTO_PATH, currentPhotoPath!!.absolutePath)
        }
        if (knitting != null) {
            outState!!.putLong(KNITTING_ID, knitting!!.id)
        }
        super.onSaveInstanceState(outState)
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        if (knitting != null) {
            if (isVisibleToUser) {
                // the fragment became visible because the user selected it in the view pager
                // get current knitting from database
                knitting = KnittingsDataSource.getInstance(activity).getKnitting(knitting!!.id)
            } else {
                // the fragment became invisible because the user selected another tab in the view pager
                // save current knitting to database
                KnittingsDataSource.getInstance(activity).updateKnitting(knitting!!)
            }
        }
    }

    companion object : AnkoLogger {

        private val CURRENT_PHOTO_PATH = "current_photo_path"
        private val KNITTING_ID = "knitting_id"
        private val REQUEST_IMAGE_CAPTURE = 0
        private val REQUEST_IMAGE_IMPORT = 1

        fun newInstance(knitting: Knitting): PhotoGalleryFragment {
            val fragment = PhotoGalleryFragment()
            val args = Bundle()
            args.putLong(KNITTING_ID, knitting.id)
            fragment.arguments = args
            debug("Created new PhotoGalleryFragment with knitting id: " + knitting.id)
            return fragment
        }
    }
}
